various python packages

krafczyk opened the pull request:
This branch contains a version of the package py-tornado which exposes the problem with the when clauses discussed in #2310.

davydden commented on var/spack/repos/builtin/packages/py-tornado/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('py-certifi')
+    depends_on('py-backports-abc', when='^python@:2')
+    depends_on('py-singledispatch', when='^python@:2')
+    depends_on('py-futures', when='+futures^python@:2')
```
i am not sure, but the `:2` could be `[...,2)`, i.e. not including `2`.  whereas `@2:` is including 2, i.e. `[2,...)`.

adamjstewart replied:
Ranges are inclusive, so this is correct.

davydden replied:
ok, did not know that.

davydden commented on var/spack/repos/builtin/packages/py-futures/package.py:
```diff
@@ -0,0 +1,57 @@
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+# This is a template package file for Spack.  We've put "FIXME"
```
i guess you forgot to remove these

davydden commented on var/spack/repos/builtin/packages/py-futures/package.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class PyFutures(Package):
+    """FIXME: Put a proper description of your package here."""
```
and modify docstring

davydden commented on var/spack/repos/builtin/packages/py-futures/package.py:
```diff
@@ -0,0 +1,57 @@
+class PyFutures(Package):
+    """FIXME: Put a proper description of your package here."""
+
+    # FIXME: Add a proper url for your package's homepage here.
```
and add an url

davydden commented on var/spack/repos/builtin/packages/py-monotonic/package.py:
```diff
@@ -0,0 +1,57 @@
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+# This is a template package file for Spack.  We've put "FIXME"
```
same here.

davydden commented on var/spack/repos/builtin/packages/py-monotonic/package.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class PyMonotonic(Package):
+    """FIXME: Put a proper description of your package here."""
```
here

davydden commented on var/spack/repos/builtin/packages/py-monotonic/package.py:
```diff
@@ -0,0 +1,57 @@
+class PyMonotonic(Package):
+    """FIXME: Put a proper description of your package here."""
+
+    # FIXME: Add a proper url for your package's homepage here.
```
and here

davydden commented on var/spack/repos/builtin/packages/py-singledispatch/package.py:
```diff
@@ -0,0 +1,58 @@
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+#
+# This is a template package file for Spack.  We've put "FIXME"
```
and here

davydden commented on var/spack/repos/builtin/packages/py-singledispatch/package.py:
```diff
@@ -0,0 +1,58 @@
+
+
+class PySingledispatch(Package):
+    """FIXME: Put a proper description of your package here."""
```
here

davydden commented on var/spack/repos/builtin/packages/py-singledispatch/package.py:
```diff
@@ -0,0 +1,58 @@
+class PySingledispatch(Package):
+    """FIXME: Put a proper description of your package here."""
+
+    # FIXME: Add a proper url for your package's homepage here.
```
here

davydden commented:
i edited the title, feel free to adjust it if needed.

adamjstewart commented on var/spack/repos/builtin/packages/py-jsonschema/package.py:
```diff
@@ -0,0 +1,43 @@
+    extends('python')
+    depends_on('py-setuptools', type='build')
+    depends_on('py-vcversioner')
+    depends_on('py-functools32')
```
There are a couple other optional dependencies that could be added, but I'm not sure if it really matters:
```python
extras_require = {                                                              
    "format" : ["rfc3987", "strict-rfc3339", "webcolors"],                      
    ":python_version=='2.6'": ["argparse", "repoze.lru"],                       
    ":python_version=='2.7'": ["functools32"],                                  
}
```

adamjstewart commented on var/spack/repos/builtin/packages/py-singledispatch/package.py:
```diff
@@ -0,0 +1,58 @@
+    extends('python')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-six')
```
`py-ordereddict` is also required for Python versions less than 2.7.

krafczyk replied:
Added in a111742c2.

adamjstewart commented on var/spack/repos/builtin/packages/py-tornado/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('py-futures', when='+futures^python@:2')
+    depends_on('py-monotonic', when='+monotonic^python@:2')
+    depends_on('py-pycurl', when='+curl')
+    depends_on('py-twisted', when='+twisted')
```
The `setup.py` doesn't mention integrating with `py-futures` through `py-twisted`. Where do they come from?

krafczyk replied:
I added those since they're listed as optional dependencies in the Arch py-tornado package. I'm looking into it.

adamjstewart commented on var/spack/repos/builtin/packages/py-zmq/package.py:
```diff
@@ -0,0 +1,42 @@
+
+    extends('python')
+    depends_on('py-setuptools', type='build')
+    depends_on('py-cython')
```
Requires `py-cython@0.16:`, `py-py`, and `py-cffi`

krafczyk replied:
Now done in 951c82edd02

adamjstewart commented on var/spack/repos/builtin/packages/py-futures/package.py:
```diff
@@ -0,0 +1,57 @@
+
+    # FIXME: Add a proper url for your package's homepage here.
+    homepage = "http://www.example.com"
+    url      = "https://pypi.python.org/packages/55/db/97c1ca37edab586a1ae03d6892b6633d8eaa23b23ac40c7e5bbc55423c78/futures-3.0.5.tar.gz"
```
For url's like this where the PyPi link contains the hash of the package, things like `spack versions` and `spack checksum` won't work. Can you try out the pypi.io link format suggested in https://github.com/LLNL/spack/issues/1939#issuecomment-252160649?

krafczyk commented:
@adamjstewart @davydden I will work on all of these issues, however the purpose of this branch is to show the `when` clause errors. my branch `features/numerous_py_packages`. I'm creating a pull request for that branch as well.

davydden commented:
@krafczyk i would say it's better to keep everything in one branch and since @adamjstewart already commented a lot, it would be easier if you just push your changes here. I don't see a reason for another branch with the same content.

davydden commented:
@krafczyk : looking at commits, something is wrong here (same commits have different hashes and are duplicated). I suggest you (i) reset your local branch to 7d18593; (ii) rebase on top of current develop in Spack (don't forget ot fetch it or pull it if it's your local copy); (iii) cherry-pick 3f0a6e1 and 3bf1d9b. 

This should give you a clean local branch. If all looks good, you can then `git push --force` to your remote on GitHub. If something went south, you can always `git reset --hard` to the remote branch on Github.

krafczyk commented:
@davydden : I've done what you suggested. I'll pay closer attention next time.

davydden commented:
@krafczyk thanks, looks nice now 👍

krafczyk commented on var/spack/repos/builtin/packages/py-tornado/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('py-setuptools', type='build')
+    depends_on('py-certifi')
+    depends_on('py-backports-abc', when='^python@:2')
+    depends_on('py-singledispatch', when='^python@:2')
```
Corrected this in d5ed4372368, Also added backports.ssl_match_hostname.

krafczyk commented:
@adamjstewart, @davydden, @citibeth, @jppelteret: I've nearly finished adding everything necessary for the jupyter-notebook to work. Now I'm running into a strange problem having to do with npm.

Since I'm not that familiar with npm, perhaps one of you can help me out here.

When I try to install py-jupyter_notebook from 0d36e3d, I get the output located at this pastebin: http://pastebin.com/rt1Ghv4Z

The log file from the build is at this pastebin: http://pastebin.com/fZR4a8z1

So, while the correct node is used along with the correct npm, I'm getting weird access errors, and npm is trying to write to my home directory. I would think ideally that anything npm installs ought to belong to the py-jupyter_notebook prefix. I'm not sure how to make that happen though.

citibeth commented:
Sigh, yet another package manager with its own ideas of how to do things
that breaks spack.  You need to figure out how to install jupyter without
npm.   Or trick npm by setting a fake home directory.
On Nov 17, 2016 4:47 PM, "Matthew Scott Krafczyk" <notifications@github.com>
wrote:

> @adamjstewart <https://github.com/adamjstewart>, @davydden
> <https://github.com/davydden>, @citibeth <https://github.com/citibeth>,
> @jppelteret <https://github.com/jppelteret>: I've nearly finished adding
> everything necessary for the jupyter-notebook to work. Now I'm running into
> a strange problem having to do with npm.
>
> Since I'm not that familiar with npm, perhaps one of you can help me out
> here.
>
> When I try to install py-jupyter_notebook from 0d36e3d
> <https://github.com/LLNL/spack/commit/0d36e3d8d4978335943fe0d6cb8f3b9a6b56c503>,
> I get the output located at this pastebin: http://pastebin.com/rt1Ghv4Z
>
> The log file from the build is at this pastebin:
> http://pastebin.com/fZR4a8z1
>
> So, while the correct node is used along with the correct npm, I'm getting
> weird access errors, and npm is trying to write to my home directory. I
> would think ideally that anything npm installs ought to belong to the
> py-jupyter_notebook prefix. I'm not sure how to make that happen though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/2320#issuecomment-261379931>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdwm6NjPgvy9S0g7hRhSaOHPDD0sxks5q_MtOgaJpZM4Kwek5>
> .
>

krafczyk commented:
Is there a way to set environment variables when installing or using a specific package?
If there is, is there a way to append  like what goes on in PATH?

tgamblin commented:
@krafczyk: See the [section on `setup_dependent_environment`](http://spack.readthedocs.io/en/latest/packaging_guide.html#setup-dependent-environment).  You can customize what goes in the build environment of dependents, and in the run environment of things that use a package.

davydden commented:
@krafczyk you can also set environment by `os.environ['SLEPC_DIR'] = os.getcwd()` or alike.

krafczyk commented:
Great!

tgamblin commented:
@krafczyk: note that if you modify `os.environ` inside `setup_dependent_environment`, that will succeed for modifying the build environment, because `setup_dependent_environment` runs for every dependency of a package before it is built.  However, it won't get things into a package's module file -- `run_env` does that.  `spack_env` is provided as a convenience so that the API for setting up the build environment is the same as the one for the run environment.

There is perhaps better documentation on this in the [module configuration section](http://spack.readthedocs.io/en/latest/module_file_support.html#package-file-api).

davydden commented on var/spack/repos/builtin/packages/flex/package.py:
```diff
@@ -28,11 +28,10 @@
+    url = "https://github.com/westes/flex/releases/download/v2.6.2/flex-2.6.2.tar.gz"
 
-    version('2.6.0', '5724bcffed4ebe39e9b55a9be80859ec')
-    version('2.5.39', 'e133e9ead8ec0a58d81166b461244fde')
```
why removing `2.5.39` and `2.6.0` ? `2.5.39` is need by `foam-extend`, for example. Removing versions is generally not a good idea.

btw, the old download works fine for me (i checked `2.5.39`). Also if that is not needed for `python` packages, then please create a separate PR for this change.

krafczyk replied:
Sorry about that, I've fixed up flex good and proper this time.

The reason I did that was because I was running into a segfault when building binutils and moving to the latest version of flex (2.6.2) seemed to solve the problem. In addition, flex development has moved to github so It would be good to reflect that change. The md5s have all been updated.

krafczyk replied:
binutils is part of the dependency change for yt which was my ultimate goal for using jupyter-notebook.

krafczyk replied:
@davydden, I've created #2385 to deal with this commit. I've also rebased this branch and removed the `flex` commits.

krafczyk commented:
I'd like to announce that jupyter-notebook is now working as expected for me! It took a lot of work, but now I'm able to use yt in jupyter-notebook!

About the npm problem, it turns out that npm was trying to write to the 'cache' which was set to ~/.npm. This can be overridden with the environment variable npm_config_cache, so I've set this to the npm prefix. This means that all npm downloads will go into npm's prefix. This is a kludgy fix, but it seems to work for now.

There are quite a few commits, so I'd like comments from folks.

citibeth commented:
The npm thing means that npm downloads cannot be versioned independently of
npm itself or of each other.  I suppose the truly right way to do this
(other than installing npm downloads correctly as with the rest of Spack)
would be to put them all the the Jupyter prefix.

On Fri, Nov 18, 2016 at 7:52 PM, Matthew Scott Krafczyk <
notifications@github.com> wrote:

> I'd like to announce that jupyter-notebook is now working as expected for
> me! It took a lot of work, but now I'm able to use yt in jupyter-notebook!
>
> About the npm problem, it turns out that npm was trying to write to the
> 'cache' which was set to ~/.npm. This can be overridden with the
> environment variable npm_config_cache, so I've set this to the npm prefix.
> This means that all npm downloads will go into npm's prefix. This is a
> kludgy fix, but it seems to work for now.
>
> There are quite a few commits, so I'd like comments from folks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/2320#issuecomment-261679654>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd9YZ6gUfHmpuOpeHzyv50BMLlBhSks5q_khmgaJpZM4Kwek5>
> .
>

adamjstewart commented on var/spack/repos/builtin/packages/node-js/package.py:
```diff
@@ -0,0 +1,106 @@
+        # So, we ensure that this is the case by...
+        if sys.platform == 'darwin':
+            result_which = subprocess.check_output(["which", "libtool"])
+            result_whereis = subprocess.check_output(["whereis", "libtool"])
```
The unit tests are crashing because `subprocess.check_output()` doesn't exist in Python 2.6. You'll have to find another way.

citibeth replied:
See: http://stackoverflow.com/questions/4814970/subprocess-check-output-doesnt-seem-to-exist-python-2-6-5

krafczyk replied:
12095e3 attempts to fix the CI errors by incorporating the fix from that stackoverflow page. I don't run OSX so I can't check that it provides the intended behavior.

krafczyk commented:
@tgamblin, I'm trying to set up the npm package to set the npm_config_cache environment variable to whatever package prefix requires npm at that time as suggested by @citibeth above. Let's say package B depends on package A. Is it possible to get the prefix of package B in the `setup_dependent_environment`
setup_dependent_environment function of package A?

krafczyk commented:
@tgamblin, any idea how to implement my previous comment?

krafczyk commented:
@davydden, Could you remove your review request about the `flex` thing from this PR? The changes fixing that problem were merged in #2384.

krafczyk commented:
@citibeth I've removed the `n-bower` package I created in favor of using `npm` whenever `npm` packages are needed. The `npm` package now sets the `npm_config_cache` variable to the `npm-cache` subdirectory of the dependent package's install prefix and creates it if necessary. This way all packages a specific package needs will be downloaded and installed in that package's prefix.

I suspect that should work for _most_ packages which use `npm` in some way.

Thoughts?

davydden commented on var/spack/repos/builtin/packages/py-jupyter_core/package.py:
```diff
@@ -0,0 +1,55 @@
+class PyJupyterCore(Package):
+    """Core Jupyter functionality"""
+
+    # FIXME: Add a proper url for your package's homepage here.
```
Don't need this fixme

davydden commented on var/spack/repos/builtin/packages/py-jupyter_core/package.py:
```diff
@@ -0,0 +1,55 @@
+
+    extends('python')
+
+    # FIXME: Add additional dependencies if required.
```
Same here

davydden commented on var/spack/repos/builtin/packages/py-jupyter_core/package.py:
```diff
@@ -0,0 +1,55 @@
+    depends_on('py-traitlets')
+
+    def install(self, spec, prefix):
+        # FIXME: Add logic to build and install here.
```
Here

davydden commented on var/spack/repos/builtin/packages/py-widgetsnbextension/package.py:
```diff
@@ -0,0 +1,44 @@
+class PyWidgetsnbextension(Package):
+    """IPython HTML widgets for Jupyter"""
+
+    # FIXME: Add a proper url for your package's homepage here.
```
Here

adamjstewart commented on var/spack/repos/builtin/packages/py-functools32/package.py:
```diff
@@ -0,0 +1,41 @@
+
+    version('3.2.3-2', '09f24ffd9af9f6cd0f63cb9f4e23d4b2')
+
+    extends('python', type=nolink)
```
Does this actually work? I've never seen something extend Python with a type of nolink before. Either way, it probably isn't necessary.

adamjstewart commented on var/spack/repos/builtin/packages/node-js/package.py:
```diff
@@ -0,0 +1,110 @@
+    variant('doc',             default=False, description='Compile with documentation')
+    variant('icu4c',           default=False, description='Build with support for all locales instead of just English')
+    variant('openssl',         default=True,  description='Build with Spacks OpenSSL instead of the bundled version')
+    variant('zlib',            default=True,  description='Build with Spacks zlib instead of the bundled version')
```
These probably don't need to be indented this far.

adamjstewart commented on var/spack/repos/builtin/packages/node-js/package.py:
```diff
@@ -0,0 +1,110 @@
+    variant('openssl',         default=True,  description='Build with Spacks OpenSSL instead of the bundled version')
+    variant('zlib',            default=True,  description='Build with Spacks zlib instead of the bundled version')
+
+    # depends_on('libtool',         type='build') # if sys.platform != 'darwin'
```
You can actually encode this into the `depends_on` directive:
```python
depends_on('libtool', type='build', when=sys.platform != 'darwin')
```

adamjstewart commented on var/spack/repos/builtin/packages/py-entrypoints/package.py:
```diff
@@ -0,0 +1,40 @@
+
+    def install(self, spec, prefix):
+        pip = which('pip')
+        pip('install', self.stage.archive_file, '--prefix={0}'.format(prefix))
```
Never seen a package that requires pip before, but until yesterday I had never seen a package that required conda before either. There's no setup.py, so I guess this is the only way to install it.

adamjstewart commented on var/spack/repos/builtin/packages/py-nbconvert/package.py:
```diff
@@ -0,0 +1,54 @@
+
+    version('4.2.0'  , '8bd88771cc00f575d5edcd0b5197f964')
+    version('4.1.0'  , '06655576713ba1ff7cece2b92760c187')
+    version('4.0.0'  , '9661620b1e10a7b46f314588d2d0932f')
```
Can you get rid of the spaces before these commas?

adamjstewart commented on var/spack/repos/builtin/packages/py-prompt_toolkit/package.py:
```diff
@@ -0,0 +1,43 @@
+from spack import *
+
+
+class PyPromptToolkit(Package):
```
Can you rename this from `py-prompt_toolkit` to `py-prompt-toolkit`? Just rename the directory that contains this file. You should also rebase now that #2475 is merged. We don't allow packages containing capital letters or underscores anymore.

adamjstewart commented on var/spack/repos/builtin/packages/py-terminado/package.py:
```diff
@@ -0,0 +1,43 @@
+class PyTerminado(Package):
+    """Terminals served to term.js using Tornado websockets"""
+
+    homepage = "http://www.example.com"
```
Can you give this a real homepage?

adamjstewart commented on var/spack/repos/builtin/packages/py-simplegeneric/package.py:
```diff
@@ -0,0 +1,43 @@
+    """Simple generic functions (similar to Python's own len(), 
+    pickle.dump(), etc.)"""
+
+    homepage = "http://www.example.com"
```
Can you give this a real homepage?

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_client/package.py:
```diff
@@ -0,0 +1,52 @@
+from spack import *
+
+
+class PyJupyterClient(Package):
```
Can you rename this from `py-jupyter_client` to `py-jupyter-client`? Just rename the directory containing this file.

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_client/package.py:
```diff
@@ -0,0 +1,52 @@
+    depends_on('py-setuptools', type='build')
+    depends_on('python@2.7:2.7.999,3.3:')
+    depends_on('py-traitlets')
+    depends_on('py-jupyter_core')
```
This will need to be renamed `py-jupyter-core`

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_console/package.py:
```diff
@@ -0,0 +1,51 @@
+from spack import *
+
+
+class PyJupyterConsole(Package):
```
Can you rename this package to `py-jupyter-console`?

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_console/package.py:
```diff
@@ -0,0 +1,51 @@
+
+    depends_on('py-setuptools', type='build')
+    depends_on('python@2.7:2.7.999,3.3:')
+    depends_on('py-jupyter_client')
```
Rename: `py-jupyter-client`

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_core/package.py:
```diff
@@ -0,0 +1,55 @@
+from spack import *
+
+
+class PyJupyterCore(Package):
```
Rename: `py-jupyter-core`

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_notebook/package.py:
```diff
@@ -0,0 +1,66 @@
+from spack import *
+
+
+class PyJupyterNotebook(Package):
```
Rename: `py-jupyter-notebook`

adamjstewart commented on var/spack/repos/builtin/packages/py-jupyter_notebook/package.py:
```diff
@@ -0,0 +1,66 @@
+    depends_on('py-traitlets')
+    depends_on('py-jupyter_core')
+    depends_on('py-jupyter_client')
+    depends_on('py-jupyter_console')
```
All 3 need to be renamed with dashes instead of underscores.

adamjstewart commented on var/spack/repos/builtin/packages/py-nbconvert/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('py-jinja2')
+    depends_on('py-pygments')
+    depends_on('py-traitlets')
+    depends_on('py-jupyter_core')
```
`py-jupyter-core`

adamjstewart commented on var/spack/repos/builtin/packages/py-nbconvert/package.py:
```diff
@@ -0,0 +1,54 @@
+    depends_on('py-nbformat')
+    depends_on('py-entrypoints')
+    depends_on('py-tornado')
+    depends_on('py-jupyter_client')
```
`py-jupyter-client`

adamjstewart commented on var/spack/repos/builtin/packages/py-nbformat/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on('py-ipython_genutils')
+    depends_on('py-traitlets')
+    depends_on('py-jsonschema')
+    depends_on('py-jupyter_core')
```
`py-jupyter-core`

adamjstewart commented on var/spack/repos/builtin/packages/py-widgetsnbextension/package.py:
```diff
@@ -0,0 +1,44 @@
+
+    depends_on('py-setuptools', type='build')
+    depends_on('python@2.7:2.7.999,3.3:')
+    depends_on('py-jupyter_notebook@4.2.0:')
```
`py-jupyter-notebook`

adamjstewart replied:
Looks like you missed this one.

adamjstewart commented on var/spack/repos/builtin/packages/py-ipykernel/package.py:
```diff
@@ -0,0 +1,56 @@
+    depends_on('py-traitlets@4.1.0:')
+    depends_on('py-tornado@4.0:')
+    depends_on('py-ipython@4.0:')
+    depends_on('py-jupyter_client')
```
`py-jupyter-client`

alalazo commented:
@tgamblin I'll try to have a look at it right now

alalazo commented:
@tgamblin Does pulling #2748 solve the issue with the tests?

krafczyk commented:
@davydden: I've removed the FIXME's you mentioned

@adamjstewart: I've fixed the grammar issues you raised as well as changed the package names you mentioned.

Do you guys think I should squash all these changes into one commit? as it stands there are almost 100 commits of work here.

citibeth commented:
I believe it will be squashed when it's merged.

On Thu, Jan 5, 2017 at 12:30 PM, Matthew Scott Krafczyk <
notifications@github.com> wrote:

> @davydden <https://github.com/davydden>: I've removed the FIXME's you
> mentioned
>
> @adamjstewart <https://github.com/adamjstewart>: I've fixed the grammar
> issues you raised as well as changed the package names you mentioned.
>
> Do you guys think I should squash all these changes into one commit? as it
> stands there are almost 100 commits of work here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/2320#issuecomment-270704374>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd7r7bN2nExIAJyhEeIWQrX5UiU4hks5rPSiigaJpZM4Kwek5>
> .
>

adamjstewart commented on var/spack/repos/builtin/packages/py-ipython_genutils/package.py:
```diff
@@ -0,0 +1,42 @@
+##############################################################################
```
This one still needs to be renamed.

krafczyk replied:
Fixed.

adamjstewart commented on var/spack/repos/builtin/packages/node-js/package.py:
```diff
@@ -0,0 +1,110 @@
+                                            stdout=subprocess.PIPE)
+            result_which = process_pipe.communicate()[0]
+            process_pipe = subprocess.Popen(["whereis", "libtool"], 
+                                            stdout=subprocess.PIPE)
```
You ought to be able to use `which('libtool')`, but I'm fine with merging this as is.

adamjstewart commented on var/spack/repos/builtin/packages/py-widgetsnbextension/package.py:
```diff
@@ -0,0 +1,43 @@
+
+    depends_on('py-setuptools', type='build')
+    depends_on('python@2.7:2.7.999,3.3:')
+    depends_on('py-jupyter_notebook@4.2.0:')
```
Last one, I think.

davydden commented:
I would not squash it even during the merge as commits have different authorship

tgamblin commented:
@alalazo:
> Does pulling #2748 solve the issue with the tests?
Yep! Perfect!

tgamblin commented:
@adamjstewart: approve?

krafczyk commented:
@tgamblin, I _did not_ know that you could make a commit on my fork without me pulling it. Nifty!

tgamblin commented:
@krafczyk: There's a checkbox for this when you commit the PR -- it's the allow edits by maintainers one.

tgamblin commented:
I squashed this into two commits by @jppelteret and @krafczyk.  They now list the added packages:

```
$ spack pkg diff develop HEAD^
HEAD^:
    node-js  py-backports-abc  py-functools32  py-pycurl   py-vcversioner
    npm      py-certifi        py-jsonschema   py-tornado  py-zmq

$ spack pkg diff HEAD^ HEAD
HEAD:
    py-backports-shutil-get-terminal-size  py-nbformat
    py-backports-ssl-match-hostname        py-pathlib2
    py-entrypoints                         py-pickleshare
    py-futures                             py-pip
    py-ipykernel                           py-prompt-toolkit
    py-ipython-genutils                    py-ptyprocess
    py-ipywidgets                          py-simplegeneric
    py-jupyter-client                      py-singledispatch
    py-jupyter-console                     py-terminado
    py-jupyter-core                        py-traitlets
    py-jupyter-notebook                    py-wcwidth
    py-monotonic                           py-widgetsnbextension
```

@krafczyk: if the tests pass and you're ok with this we can merge.

krafczyk commented:
@tgamblin I'm testing that things still work for me as I expect. I'll let you know in a bit.

tgamblin commented:
@krafczyk: Sounds good.  If you need to revert, the tip before rebasing was b6edd93.  The only diff between that and the two squashed commits above is that the squashed ones are based on `develop` with @alalazo's test fix.

```console
$ git co fix/when_variants_plus_dependency 
Switched to branch 'fix/when_variants_plus_dependency'
Your branch is up-to-date with 'krafczyk/fix/when_variants_plus_dependency'.
$ git diff b6edd93
diff --git a/lib/spack/spack/test/conftest.py b/lib/spack/spack/test/conftest.py
index 11127d8..79d9018 100644
--- a/lib/spack/spack/test/conftest.py
+++ b/lib/spack/spack/test/conftest.py
@@ -122,7 +122,6 @@ def builtin_mock(repo_path):
 def refresh_builtin_mock(builtin_mock, repo_path):
     """Refreshes the state of spack.repo"""
     # Get back the real repository
-    spack.repo.swap(builtin_mock.real)
     mock_repo = copy.deepcopy(repo_path)
     spack.repo.swap(mock_repo)
     return builtin_mock
```

krafczyk commented:
@tgamblin: Everything looks good to me. You can merge when the tests are finished.

adamjstewart commented:
@tgamblin @alalazo Any idea why the macOS tests take so much longer?

alalazo commented:
@adamjstewart It's the set-up phase that seems to take a lot (e.g `brew update` & Co.).

krafczyk commented:
@tgamblin tests are done, ready to merge

tgamblin commented:
@krafczyk: merged!  Thanks a lot to you and @jppelteret!
